util/patricia_trie/src/lookup.rs

```diff
@@ -16,27 +16,35 @@
 
 //! Trie lookup via HashDB.
 
-use hashdb::HashDB;
+use hashdb::{HashDB, Hasher};
 use nibbleslice::NibbleSlice;
-use ethereum_types::H256;
-
-use super::{TrieError, Query};
+use rlp::Decodable;
 use super::node::Node;
+use node_codec::NodeCodec;
+use super::{TrieError, Query};
+use std::marker::PhantomData;
 
 /// Trie lookup helper object.
-pub struct Lookup<'a, Q: Query> {
+pub struct Lookup<'a, H: Hasher + 'a, C: NodeCodec<H>, Q: Query<H>> {
 	/// database to query from.
-	pub db: &'a HashDB,
+	pub db: &'a HashDB<H>,
 	/// Query object to record nodes and transform data.
 	pub query: Q,
 	/// Hash to start at
-	pub hash: H256,
+	pub hash: H::Out,
+	pub marker: PhantomData<C>, // TODO: probably not needed when all is said and done? When Query is made generic?
```

Comment:
`pub codec: &'a C` would probably be fine

Reply:
I can't get that to work. I need:

```rust
pub struct Lookup<'a, H: Hasher + 'a, C: NodeCodec<H> + 'a, Q: Query<H>> {
	pub db: &'a HashDB<H>,
	pub query: Q,
	pub hash: H::Out,
	pub codec: &'a C,
}
```

Reply:
assuming the trie itself has a node-codec member then it can simply borrow it